### PR TITLE
Include App Nginx Plus

### PR DESCRIPTION
**Pull request by nitishkumar71**
Add support of [NGINX Ingress](https://docs.nginx.com/nginx-ingress-controller/overview/)
## Description

1. Included a new file to add NGINX ingress as an application.
2. Included into install command list
## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
#234 

## How Has This Been Tested?

1. run `make test`, all existing test cases passed
2. created build using `make build`
### Install NGINX inc OSS
1. installed NGINX inc OSS Ingress using `arkade install nginx-inc`
2. Deploy Echo server `kubectl apply -f https://bit.ly/echo-service`
3. Create Ingress class
```
echo "
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: demo
  annotations:
    kubernetes.io/ingress.class: nginx
spec:
  rules:
  - host: <replace with domain name>
    http:
      paths:
      - path: /foo
        backend:
          serviceName: echo
          servicePort: 80
" | kubectl apply -f -
```
4. Ping exposed endpoint

### Install Nginx Plus
1. Build ingress controller image and push to private repository. Follow the steps from official [Doc](https://docs.nginx.com/nginx-ingress-controller/installation/building-ingress-controller-image/#building-the-image-and-pushing-it-to-the-private-registry)
2. installed NGINX inc OSS Ingress using `arkade install nginx-inc --controller-image=<replace with your private ingress controller image built above> --plus true`
3. Deploy Echo server `kubectl apply -f https://bit.ly/echo-service`
4. Create Ingress class
```
echo "
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: demo
  annotations:
    kubernetes.io/ingress.class: nginx
spec:
  rules:
  - host: <replace with domain name>
    http:
      paths:
      - path: /foo
        backend:
          serviceName: echo
          servicePort: 80
" | kubectl apply -f -
```
5. Ping exposed endpoint
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

- [ ] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

**Comment by alexellis** [comment]
Hi @nitishkumar71 this looks good. I'm surprised that they used the same ingressClass `nginx` as the community project `ingress-nginx`, but if that's the case, fair enough.

What's left on the PR, it says WIP? 

Alex

**Comment by nitishkumar71** [comment]
@alexellis 
I had not updated steps to install NGINX Plus, now it's done

**Comment by alexellis** [comment]
There is a conflict on a file now, could you rebase please?

What other flags do you think users may want to expose from the chart?

```
--set controller.nginxplus=true
```

As `--plus` ?

```
controller.replicaCount
```
as: `--replicas`

```
prometheus.create
````

as : `--prometheus` true/false (default false)

**Comment by nitishkumar71** [comment]
> There is a conflict on a file now, could you rebase please?
> 
> What other flags do you think users may want to expose from the chart?
> 
> ```
> --set controller.nginxplus=true
> ```
> 
> As `--plus` ?
> 
> ```
> controller.replicaCount
> ```
> 
> as: `--replicas`
> 
> ```
> prometheus.create
> ```
> 
> as : `--prometheus` true/false (default false)

Sure, will rebase it and add the following flags along with 

`prometheus.port` as `--prometheus-port` with default 9113
`controller.kind` as `--kind` with default deployment

**Comment by alexellis** [comment]
SGTM.

File: cmd/apps/nginx_inc_app.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		clientArch, clientOS := env.GetClientArch()
+
+		fmt.Printf("Client: %s, %s\n", clientArch, clientOS)
+		log.Printf("User dir established as: %s\n", userPath)
+
+		os.Setenv("HELM_HOME", path.Join(userPath, ".helm"))
+
+		overrides := map[string]string{}
+
+		nginxPlus, _ := command.Flags().GetBool("plus")
```
Comment (alexellis) [inline cmd/apps/nginx_inc_app.go:61]: I know that not all the apps do this, but for each please add:

```
nginxPlus, err := command.Flags().GetBool("plus")
if err!=nil {
   return err
}
```

This is required to pick up errors.

Once added here, happy for you to retrofit elsewhere too.

File: cmd/apps/nginx_inc_app.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		overrides["controller.nginxplus"] = fmt.Sprintf("%t", nginxPlus)
```
Comment (alexellis) [inline cmd/apps/nginx_inc_app.go:62]: There is a method in strconv that can do this.

File: cmd/apps/nginx_inc_app.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+		controllerImage, _ := command.Flags().GetString("controller-image")
+		if controllerImage != "" {
```
Comment (alexellis) [inline cmd/apps/nginx_inc_app.go:65]: Add to this: `command.Flags().HasChanged()`

File: cmd/apps/nginx_inc_app.go (hunk 1 above)
Comment (nitishkumar71) [reply cmd/apps/nginx_inc_app.go]: Thanks for mentioning it. It's done

File: cmd/apps/nginx_inc_app.go (hunk 2 above)
Comment (nitishkumar71) [reply cmd/apps/nginx_inc_app.go]: Done

File: cmd/apps/nginx_inc_app.go (hunk 3 above)
Comment (nitishkumar71) [reply cmd/apps/nginx_inc_app.go]: done

File: cmd/apps/nginx_inc_app.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+			overrides["controller.image.repository"] = controllerImage
+		}
+
+		replicaCount, err := command.Flags().GetInt("replicas")
+		if err != nil {
+			return fmt.Errorf("error with --replicas usage: %s", err)
+		}
+		overrides["controller.replicaCount"] = fmt.Sprintf("%d", replicaCount)
+
+		prometheusCreate, _ := command.Flags().GetBool("prometheus")
```
Comment (alexellis) [inline cmd/apps/nginx_inc_app.go:82]: This is missing the assignment into `err` on this line.

**Review by alexellis — CHANGES_REQUESTED** [review]
One more change required please.

File: cmd/apps/nginx_inc_app.go (hunk 4 above)
Comment (nitishkumar71) [reply cmd/apps/nginx_inc_app.go]: Done

**Comment by nitishkumar71** [comment]
> One more change required please.
Apologies, don't know how I missed that

**Comment by alexellis** [comment]
Sorry for missing your request for another review @nitishkumar71 - I don't think I got a notification.

Thank you for the contribution.

**Review by alexellis — APPROVED** [review]
Approved

**Pull request merged**